### ISSUE TITLE
USE_PTHREADS freezes Firefox Nightly 42.0a1

### ISSUE THREAD

**Issue by Kagami** [kripken/emscripten#3636]
Hi. I'm trying to make simple threads example to work with Emscripten's `-s USE_PTHREADS=1` option but unfortunately it doesn't work with latest stable Emscripten and Firefox Nightly. Probably I'm doing something stupidly wrong, sorry for that.

Here is simple pthread example:

```c
$ cat 1.c 
#include <stdio.h>
#include <pthread.h>

void *thread_fn(void *arg) {
    printf("Thread done\n");
    return NULL;
}

int main(void)
{
    pthread_t thread1, thread2;
    pthread_create(&thread1, NULL, thread_fn, NULL);
    pthread_create(&thread2, NULL, thread_fn, NULL);
    pthread_join(thread1, NULL);
    pthread_join(thread2, NULL);
    printf("Done\n");
    return 0;
}
```

I'm compiling it with the following options to `index.html`:

```
$ emcc -s USE_PTHREADS=1 1.c -o index.html
warning: unresolved symbol: posix_spawn
warning: unresolved symbol: posix_spawn_file_actions_adddup2
warning: unresolved symbol: posix_spawn_file_actions_init
warning: unresolved symbol: posix_spawn_file_actions_destroy
```

When I open the resulting page in Firefox, it freezes the window. If I stop the script, I got the following text at page's console:

```
Preallocating 1 workers for a pthread spawn pool.
Preallocating 1 workers for a pthread spawn pool.
Thread done
Thread done
```

and this in the browser's console:

```
pre-main prep time: 6 ms t:1245:13

Preallocating 1 workers for a pthread spawn pool. t:1233:13
Error: Script terminated by timeout at:
_emscripten_futex_wait@http://127.0.0.1:4444/t/index.js:5479:17
_pthread_join@http://127.0.0.1:4444/t/index.js:6862:9
_main@http://127.0.0.1:4444/t/index.js:8698:3
asm._main@http://127.0.0.1:4444/t/index.js:34880:8
callMain@http://127.0.0.1:4444/t/index.js:35041:15
doRun@http://127.0.0.1:4444/t/index.js:35100:42
run/<@http://127.0.0.1:4444/t/index.js:35111:7
 index.js:5479:17
Thread done
```

If I compile the code with pool size option:

```
emcc -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=2 1.c -o index.html
```

I got the following error at the page instead:

```
Preallocating 2 workers for a pthread spawn pool.
Assertion failed: you need to wait for the runtime to be ready (e.g. wait for main() to be called)
```

I tried to compile `test_pthread_join.cpp` from the Emscripten's test suite but got the same result.

```
emcc -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=-1 -s PTHREAD_HINT_NUM_CORES=-1 1.c -o index.html
```

gives me Thread setup popup on page load but after that I got errors in console and can't see the threads output:

```
ReferenceError: _emscripten_force_num_logical_cores is not defined index.js:1631:5
still waiting on run dependencies: t:1245:13
dependency: pthreads_querycores t:1245:13
(end of list) t:1245:13
still waiting on run dependencies: t:1245:13
dependency: pthreads_querycores t:1245:13
(end of list) t:1245:13
still waiting on run dependencies: t:1245:13
dependency: pthreads_querycores t:1245:13
(end of list) t:1245:13
```

Am I doing something wrong? I'm using `emscripten/tag-1.34.3` and `Firefox 42.0a1 (2015-07-22)`.

**Comment by kripken** [comment]
The safest thing for now is the see what is in the test suite, as nothing else is guaranteed to work.

I fixed the assertion error, it was a memory allocation issue. Also added testing. (It wasn't noticed because the test suite optimizes builds, and those disable assertions.)

It looks like all the tests allocate `PTHREAD_POOL_SIZE` to 2 or 8 or such, and without that, most tests fail. I think @juj found an issue in the browser where if the workers aren't already preallocated, it hangs, so it could be that.

**Comment by Kagami** [comment]
```
emcc -O1 -s USE_PTHREADS=1 -s PTHREAD_POOL_SIZE=2 1.c -o index.html
```

works fine. Thanks for the fast answer and fix!

**Comment by kripken** [comment]
Actually let's keep this open. I hope @juj can take a look at the hang/freeze aspect here, just to make sure it's the thing he is already aware of and not something else.

**Comment by juj** [comment]
Indeed, it was the `-s PTHREAD_POOL_SIZE=2`. Also, the (false positive) unresolved symbol warnings were resolved in the recent pull request #3819, so closing this now as resolved.
